File: src/test/java/me/tsukanov/counter/domain/impl/IntegerCounterTest.java
package me.tsukanov.counter.domain.impl;

import static org.junit.jupiter.api.Assertions.assertEquals;

import me.tsukanov.counter.domain.Counter;
import me.tsukanov.counter.domain.IntegerCounter;
import me.tsukanov.counter.domain.exception.CounterException;
import org.junit.jupiter.api.Test;

class IntegerCounterTest {
  private static final String DEFAULT_NAME = "Test name";

  @Test
  void nameSetting_worksAsExpected() throws CounterException {
    final Counter counter = new IntegerCounter(DEFAULT_NAME);
    assertEquals(DEFAULT_NAME, counter.getName());

    final String newName = "New name";
    counter.setName(newName);
    assertEquals(newName, counter.getName());
  }

  @Test
  void valueSetting() throws Exception {
    final IntegerCounter counter = new IntegerCounter(DEFAULT_NAME, 0);
    assertEquals(0, counter.getValue());

    counter.setValue(42);
    assertEquals(42, counter.getValue());

    counter.setValue(IntegerCounter.MAX_VALUE);
    assertEquals(999999999, counter.getValue());

    counter.setValue(IntegerCounter.MIN_VALUE);
    assertEquals(-99999999, counter.getValue());
  }

  @Test
  void increment() throws Exception {
    final IntegerCounter counter = new IntegerCounter(DEFAULT_NAME, 0);
    assertEquals(0, counter.getValue());

    counter.increment();
    assertEquals(1, counter.getValue());

    counter.increment();
    counter.increment();
    assertEquals(3, counter.getValue());
  }

  @Test
  void decrement() throws Exception {
    final IntegerCounter counter = new IntegerCounter(DEFAULT_NAME, 0);
    assertEquals(0, counter.getValue());

    counter.decrement();
    assertEquals(-1, counter.getValue());

    counter.decrement();
    counter.decrement();
    assertEquals(-3, counter.getValue());
  }
}

File: app/src/test/java/me/tsukanov/counter/domain/impl/IntegerCounterTest.kt
package me.tsukanov.counter.domain.impl

import me.tsukanov.counter.domain.IntegerCounter
import me.tsukanov.counter.domain.exception.CounterException
import org.junit.Assert
import org.junit.Test

class IntegerCounterTest {

    companion object {
        private const val DEFAULT_NAME = "Test name"
    }

    @Test
    @Throws(CounterException::class)
    fun nameSetting_worksAsExpected() {
        val counter = IntegerCounter(DEFAULT_NAME)
        Assert.assertEquals(DEFAULT_NAME, counter.name)

        val newName = "New name"
        counter.name = newName
        Assert.assertEquals(newName, counter.name)
    }

    @Test
    @Throws(Exception::class)
    fun valueSetting() {
        val counter = IntegerCounter(DEFAULT_NAME, 0)
        Assert.assertEquals(0, counter.value.toLong())

        counter.value = 42
        Assert.assertEquals(42, counter.value.toLong())

        counter.value = IntegerCounter.MAX_VALUE
        Assert.assertEquals(999999999, counter.value.toLong())

        counter.value = IntegerCounter.MIN_VALUE
        Assert.assertEquals(-99999999, counter.value.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun increment() {
        val counter = IntegerCounter(DEFAULT_NAME, 0)
        Assert.assertEquals(0, counter.value.toLong())

        counter.increment()
        Assert.assertEquals(1, counter.value.toLong())

        counter.increment()
        counter.increment()
        Assert.assertEquals(3, counter.value.toLong())
    }

    @Test
    @Throws(Exception::class)
    fun decrement() {
        val counter = IntegerCounter(DEFAULT_NAME, 0)
        Assert.assertEquals(0, counter.value.toLong())

        counter.decrement()
        Assert.assertEquals(-1, counter.value.toLong())

        counter.decrement()
        counter.decrement()
        Assert.assertEquals(-3, counter.value.toLong())
    }

}
